### PR TITLE
Add ´?´ handling in OSC 4 and correctly query modified colors

### LINKED ISSUE #5542: problems with OSC 4 and 104
[Documentation claims that OSC 4 and OSC 104 are supported](https://github.com/alacritty/alacritty/blob/master/docs/escape_support.md). However, testing reveals some issues.

If the OSC is fully supported, we should be able to query the state of colors. Here's a working query using OSC 11:
```
> echo -ne '\e]11;?\a'; cat
^[]11;rgb:ffff/ffff/ffff^G^C⏎  
```
That tells me I have a white background.

Attempting to learn the state of a color in my color palette using OSC 4 does not work, however:
```
> echo -ne '\e]4;1;?\a'; cat
^C⏎ 
```
No information is returned. And in the log, we see:
> [2021-10-13 11:37:43.348998800] [DEBUG] [alacritty_terminal] [unhandled osc_dispatch]: [['4',],['1',],['?',],] at line 947

Strangely, we are able to change the colors with OSC 4:
```
> printf '\e]4;3;#490101\a'
```
That sets my dark yellow to a dark red, easy to see with e.g. `neofetch`.

However, attempting to undo this with OSC 104 does not work:
```
> printf '\e]104;\a'
```
My yellow remains red. The log reports:
> [2021-10-13 11:45:18.472954361] [DEBUG] [alacritty_terminal] [unhandled osc_dispatch]: [['1','0','4',],[],] at line 947

Using a different ending has no effect on results:
```
> echo -ne '\e]11;?\e\\'; cat
^[]11;rgb:ffff/ffff/ffff^[\^C⏎     
> echo -ne '\e]4;1;?\e\\'; cat
^C⏎
```

I should be able to query my color palette with OSC 4, and reset my color palette using OSC 104. Currently I can only use OSC 4 to change colors. The above commands work fine in other shells such as iTerm2 or Foot. Similar commands for OSC 10, 11, 110, and 111 all work in Alacritty as expected.

### System

OS: macOS Catalina 10.15.7
Version: alacritty 0.9.0 (fed349a)

OS: Manjaro ARM Linux
Version: alacritty 0.10.0-dev (b6e05d2d)
DE: Wayland, sway

### Logs

> alacritty -vv
Created log file at "/tmp/Alacritty-25654.log"
[2021-10-13 11:36:58.858691035] [INFO ] [alacritty] Welcome to Alacritty
[2021-10-13 11:36:58.858972769] [INFO ] [alacritty] Configuration files loaded from:
                                                      "/home/nelson/.config/alacritty/alacritty.yml"
[2021-10-13 11:36:58.900859460] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[2021-10-13 11:36:58.910606092] [DEBUG] [alacritty] Estimated DPR: 1
[2021-10-13 11:36:58.910720127] [DEBUG] [alacritty] Estimated window size: None
[2021-10-13 11:36:58.910755125] [DEBUG] [alacritty] Estimated cell size: 9 x 20
[2021-10-13 11:36:59.019037732] [INFO ] [alacritty] Device pixel ratio: 1
[2021-10-13 11:36:59.049712562] [INFO ] [alacritty] Initializing glyph cache...
[2021-10-13 11:36:59.054074182] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[2021-10-13 11:36:59.063504963] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[2021-10-13 11:36:59.073139607] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Italic, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[2021-10-13 11:36:59.082763169] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold Italic, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[2021-10-13 11:36:59.430697721] [INFO ] [alacritty] ... finished initializing glyph cache in 0.380882498s
[2021-10-13 11:36:59.430961080] [INFO ] [alacritty] Cell size: 9 x 20
[2021-10-13 11:36:59.431031076] [INFO ] [alacritty] Padding: 4 x 10
[2021-10-13 11:36:59.431070157] [INFO ] [alacritty] Width: 800, Height: 600
[2021-10-13 11:36:59.434250888] [INFO ] [alacritty] PTY dimensions: 29 x 88
[2021-10-13 11:36:59.452195504] [INFO ] [alacritty] Initialisation complete
[2021-10-13 11:36:59.481600616] [DEBUG] [alacritty_terminal] New num_cols is 105 and num_lines is 51
[2021-10-13 11:36:59.490114743] [INFO ] [alacritty] Padding: 5 x 7
[2021-10-13 11:36:59.490217112] [INFO ] [alacritty] Width: 956, Height: 1035
[2021-10-13 11:36:59.645425319] [DEBUG] [alacritty_terminal] [unhandled osc_dispatch]: [['5','0',],['c','o','l','o','r','s','=','l','i','g','h','t',],] at line 947
[2021-10-13 11:37:05.464752222] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[2021-10-13 11:37:43.348998800] [DEBUG] [alacritty_terminal] [unhandled osc_dispatch]: [['4',],['1',],['?',],] at line 947
[2021-10-13 11:45:18.472954361] [DEBUG] [alacritty_terminal] [unhandled osc_dispatch]: [['1','0','4',],[],] at line 947

### PR DESCRIPTION
**Pull request by DaftMouse**
Related #5542

- Handle `?` in OSC 4:
```
$ echo -ne '\e]4;1;?\a'; cat
^[]1;rgb:cccc/2424/1d1d^G
```
- Modify handling of `TerminalEvent::ColorRequest` to correcly report modified colors
- Add some tests for OSC 4  parsing

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -15,6 +15,12 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 - The `--help` output was reworked with a new colorful syntax
 
+### Fixed
+
+- OSC 104 not reseting all colors when used without parameters
```
Comment (chrisduerr) [inline CHANGELOG.md:20]: Could we move this change to a separate PR? The two changes are completely unrelated.

File: alacritty/src/event.rs [hunk 2]
```diff
@@ -1074,7 +1074,10 @@ impl input::Processor<EventProxy, ActionContext<'_, Notifier, EventProxy>> {
                         self.ctx.write_to_pty(text.into_bytes());
                     },
                     TerminalEvent::ColorRequest(index, format) => {
-                        let text = format(self.ctx.display.colors[index]);
+                        let text = format(
+                            self.ctx.terminal().colors()[index]
+                                .unwrap_or(self.ctx.display.colors[index]),
+                        );
                         self.ctx.write_to_pty(text.into_bytes());
```
Comment (chrisduerr) [inline alacritty/src/event.rs:1081]:
```suggestion
                        let color = self.ctx.terminal().colors()[index]
                            .unwrap_or(self.ctx.display.colors[index]);
                        self.ctx.write_to_pty(format(color).into_bytes());
```

File: alacritty_terminal/src/ansi.rs [hunk 3]
```diff
@@ -974,13 +974,21 @@ where
             b"4" => {
                 if params.len() > 1 && params.len() % 2 != 0 {
                     for chunk in params[1..].chunks(2) {
-                        let index = parse_number(chunk[0]);
-                        let color = xparse_color(chunk[1]);
-                        if let (Some(i), Some(c)) = (index, color) {
-                            self.handler.set_color(i as usize, c);
-                            return;
+                        if let Some(index) = parse_number(chunk[0]) {
```
Comment (chrisduerr) [inline alacritty_terminal/src/ansi.rs:977]: Really not a fan of the additional nesting here, especially because it already was pretty bad.

File: alacritty_terminal/src/ansi.rs [hunk 4]
```diff
@@ -974,13 +974,21 @@ where
...
+                                    index,
+                                    index as usize,
+                                    terminator,
+                                );
+                            } else {
+                                unhandled(params);
+                            }
                         }
                     }
+                    return;
```
Comment (chrisduerr) [inline alacritty_terminal/src/ansi.rs:991]: Why was this `return` added? Is it just to avoid calling the `unhandled`? The branching of this particular match arm seems far too complicated I'd say.

File: alacritty_terminal/src/ansi.rs [hunk 5]
```diff
@@ -1053,7 +1061,7 @@ where
             // Reset color index.
             b"104" => {
                 // Reset all color indexes when no parameters are given.
-                if params.len() == 1 {
+                if params.len() > 1 && params[1].is_empty() {
```
Comment (chrisduerr) [inline alacritty_terminal/src/ansi.rs:1064]: Looks like you've just changed it from one incorrect implementation to another here. `OSC 104 ST` and `OSC 104 ; ST` should **both** work. At least I'm pretty sure that's what XTerm does.

**Review by chrisduerr — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (DaftMouse) [reply CHANGELOG.md:20]: I put it here because they're both from the same GH issue but I can make a separate PR for it.

Comment (chrisduerr) [reply CHANGELOG.md:20]: Github issue doesn't matter, just open a new PR for every separate feature please.

File: alacritty_terminal/src/ansi.rs [hunk 6]
```diff
@@ -972,17 +972,27 @@ where
...
+                    return;
+                }
+
+                for chunk in params[1..].chunks(2) {
+                    let index = if let Some(i) = parse_number(chunk[0]) {
+                        i
+                    } else {
+                        unhandled(params);
+                        continue;
+                    };
```
Comment (chrisduerr) [inline alacritty_terminal/src/ansi.rs:986]: This should be a match.

**Review by chrisduerr — CHANGES_REQUESTED** [review]
Looks better already, thanks.

File: CHANGELOG.md (hunk 1 above)
Comment (DaftMouse) [reply CHANGELOG.md]: Took me a while to get git to do what I wanted but the OSC 104 related stuff from this branch and will open another PR for that after this one

File: alacritty_terminal/src/ansi.rs [hunk 7]
```diff
@@ -1724,4 +1741,18 @@ mod tests {
...
+        let bytes: &[u8] = &[0x1b, b']', b'4', b';', b'0', b';', b'#', b'f', b'f', b'f', 0x1b];
+
+        let mut parser = Processor::new();
+        let mut handler = MockHandler::default();
+
+        for byte in bytes {
+            parser.advance(&mut handler, *byte);
+        }
+
+        assert_eq!(handler.color, xparse_color(&[b'#', b'f', b'f', b'f']));
```
Comment (chrisduerr) [inline alacritty_terminal/src/ansi.rs:1756]: Should use the Rgb here, not call `xparse_color` to verify the test result.

**Review by chrisduerr — CHANGES_REQUESTED** [review]

**Review by chrisduerr — APPROVED** [review]
